qa/vagrant/src/main/java/org/elasticsearch/packaging/test/PackageTestCase.java

```diff
@@ -0,0 +1,176 @@
+import static org.hamcrest.CoreMatchers.notNullValue;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.containsString;
+import static org.hamcrest.Matchers.isEmptyString;
+
+import static org.hamcrest.core.Is.is;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assume.assumeThat;
+import static org.junit.Assume.assumeTrue;
+
+@FixMethodOrder(MethodSorters.NAME_ASCENDING)
+public abstract class PackageTestCase {
+
+    private static Installation installation;
+
+    protected abstract Distribution distribution();
+
+    @BeforeClass
+    public static void cleanup() {
+        installation = null;
+        cleanEverything();
+    }
+
+    @Before
+    public void onlyCompatibleDistributions() {
+        assumeTrue("only compatible distributions", distribution().packaging.compatible);
+    }
+
+    @Test
+    public void test10InstallPackage() {
+        assertRemoved(distribution());
+        installation = install(distribution());
+        assertInstalled(distribution());
+        verifyPackageInstallation(installation, distribution());
+    }
+
+    @Test
+    public void test20PluginsCommandWhenNoPlugins() {
+        assumeThat(installation, is(notNullValue()));
+
+        final Shell sh = new Shell();
+        assertThat(sh.run(installation.bin("elasticsearch-plugin") + " list").stdout, isEmptyString());
+    }
+
+    @Test
+    public void test30InstallDoesNotStartServer() {
+        assumeThat(installation, is(notNullValue()));
+
+        final Shell sh = new Shell();
+        assertThat(sh.run("ps aux").stdout, not(containsString("org.elasticsearch.bootstrap.Elasticsearch")));
+    }
+
+    @Test
+    public void test40StartServer() throws IOException {
+        assumeThat(installation, is(notNullValue()));
+
+        startElasticsearch();
+        runElasticsearchTests();
+        verifyPackageInstallation(installation, distribution()); // check startup script didn't change permissions
+    }
+
+    @Test
+    public void test50Remove() {
+        assumeThat(installation, is(notNullValue()));
+
+        remove(distribution());
+
+        // removing must stop the service
+        final Shell sh = new Shell();
+        assertThat(sh.run("ps aux").stdout, not(containsString("org.elasticsearch.bootstrap.Elasticsearch")));
+
+        if (isSystemd()) {
+
+            final int statusExitCode;
+
+            // Before version 231 systemctl returned exit code 3 for both services that were stopped, and nonexistent
+            // services [1]. In version 231 and later it returns exit code 4 for non-existent services.
+            //
+            // The exception is Centos 7, where it returns exit code 4 for non-existent services from a systemd reporting a version
+            // earlier than 231. Centos 6 does not have an /etc/os-release, but that's fine because it also doesn't use systemd.
+            //
+            // [1] https://github.com/systemd/systemd/pull/3385
```

Comment:
I did not dig too deeply into why centos behaves differently here. I figured it was worth still doing this (vs just accepting exit codes 3 or 4) because distinguishing between exists-but-stopped and nonexistent seems valuable if we're able to